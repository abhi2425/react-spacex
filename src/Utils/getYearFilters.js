export const getYearlyFilters = () => [
  {
    label: 2006,
  },
  {
    label: 2007,
  },
  {
    label: 2008,
  },
  {
    label: 2009,
  },
  {
    label: 2010,
  },
  {
    label: 2011,
  },
  {
    label: 2012,
  },
  {
    label: 2013,
  },
  {
    label: 2014,
  },
  {
    label: 2015,
  },
  {
    label: 2016,
  },
  {
    label: 2017,
  },
  {
    label: 2018,
  },
  {
    label: 2019,
  },
  {
    label: 2020,
  },
]
