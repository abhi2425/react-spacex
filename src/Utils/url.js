import axios from 'axios'

export const BASE_URL = 'https://api.spacexdata.com/v3/'
axios.defaults.baseURL = BASE_URL

export { axios }
