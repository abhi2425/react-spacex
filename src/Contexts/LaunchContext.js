import { axios } from '../Utils/url'
import React, {
  createContext,
  useCallback,
  useContext,
  useEffect,
  useMemo,
  useState,
} from 'react'
import window from 'global/window'

const LaunchContext = createContext()

export const LaunchContextProvider = ({ children }) => {
  const [allSpaceShips, setAllSpaceShips] = useState([])
  const [loading, setLoading] = useState(true)
  const [screenWidth, setScreenWidth] = useState(window.innerWidth)
  const [url, setUrl] = useState('')

  const getAllSpaceShips = useCallback(async (url = 'launches?limit=100') => {
    try {
      setLoading(true)
      const { data } = await axios.get(url)
      if (data) {
        setLoading(false)
        setAllSpaceShips(data)
      }
    } catch (error) {
      setLoading(false)
      console.log(error)
    }
  }, [])

  useEffect(() => {
    if (url) {
      getAllSpaceShips(url)
      window.history.replaceState(null, 'SpaceX Launch Program', url)
    }
  }, [url, getAllSpaceShips])

  const values = useMemo(
    () => ({
      getAllSpaceShips,
      setScreenWidth,
      setUrl,
      allSpaceShips,
      loading,
      screenWidth,
    }),
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [allSpaceShips, loading, screenWidth]
  )

  return (
    <LaunchContext.Provider value={values}>{children}</LaunchContext.Provider>
  )
}

export const useLaunchContext = () => useContext(LaunchContext)
