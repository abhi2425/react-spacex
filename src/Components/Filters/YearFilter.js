import React, { memo, useCallback } from 'react'
import { useLaunchContext } from '../../Contexts/LaunchContext'

const YearFilter = ({ label, active, setActiveFilter }) => {
  const { setUrl } = useLaunchContext()

  const yearlyFilterHandler = useCallback(() => {
    localStorage.setItem('yearFilter', JSON.stringify(label))

    const yearFilter = localStorage.getItem('yearFilter')
    const launch_success = localStorage.getItem('launch_success')
    const land_success = localStorage.getItem('land_success')
    const url = yearFilter
      ? `launches?launch_year=${yearFilter}`
      : 'launches?limit=100'

    if (launch_success && land_success) {
      setUrl(
        `${url}&launch_success=${launch_success}&land_success=${land_success}`
      )
      return
    }
    if (launch_success) {
      setUrl(`${url}&launch_success=${launch_success}`)
      return
    }
    if (land_success) {
      setUrl(
        `${url}&launch_success=${launch_success}&launch_year=${yearFilter}`
      )
      return
    }

    setUrl(url)
  }, [label, setUrl])

  return (
    <button
      className={`filter-button ${active}`}
      onClick={() => {
        setActiveFilter()
        yearlyFilterHandler()
      }}
    >
      {label}
    </button>
  )
}

export default memo(YearFilter)
