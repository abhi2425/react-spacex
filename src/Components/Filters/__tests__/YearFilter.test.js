import React from 'react'
import ReactDOM from 'react-dom'
import YearFilter from '../YearFilter'
import { render, cleanup } from '@testing-library/react'
import '@testing-library/jest-dom'
import renderer from 'react-test-renderer'
import { LaunchContextProvider } from '../../../Contexts/LaunchContext'
import { getYearlyFilters } from '../../../Utils/getYearFilters'

afterEach(cleanup)
it('should render button', () => {
  const div = document.createElement('div')
  ReactDOM.render(
    <LaunchContextProvider>
      <YearFilter active={true} label={2016} setActiveFilter={() => {}} />
    </LaunchContextProvider>,
    div
  )
  const button = getYearlyFilters().map(({ label }, index) => (
    <LaunchContextProvider key={index}>
      <YearFilter active={true} label={label} setActiveFilter={() => {}} />
    </LaunchContextProvider>
  ))
  const { getAllByTestId } = render(button)
  getAllByTestId('yearFilter').map((data) =>
    expect(data).toHaveClass('filter-button true')
  )
})

it('should match button snapshot', () => {
  const button = getYearlyFilters().map(({ label }, index) => (
    <LaunchContextProvider key={index}>
      <YearFilter active={''} label={label} setActiveFilter={() => {}} />
    </LaunchContextProvider>
  ))
  const tree = renderer.create(button)
  expect(tree).toMatchSnapshot()
})
