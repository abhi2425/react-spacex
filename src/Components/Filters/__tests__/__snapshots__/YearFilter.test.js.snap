// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match button snapshot 1`] = `
Array [
  <button
    className="filter-button "
    data-testid="yearFilter"
    onClick={[Function]}
  >
    2006
  </button>,
  <button
    className="filter-button "
    data-testid="yearFilter"
    onClick={[Function]}
  >
    2007
  </button>,
  <button
    className="filter-button "
    data-testid="yearFilter"
    onClick={[Function]}
  >
    2008
  </button>,
  <button
    className="filter-button "
    data-testid="yearFilter"
    onClick={[Function]}
  >
    2009
  </button>,
  <button
    className="filter-button "
    data-testid="yearFilter"
    onClick={[Function]}
  >
    2010
  </button>,
  <button
    className="filter-button "
    data-testid="yearFilter"
    onClick={[Function]}
  >
    2011
  </button>,
  <button
    className="filter-button "
    data-testid="yearFilter"
    onClick={[Function]}
  >
    2012
  </button>,
  <button
    className="filter-button "
    data-testid="yearFilter"
    onClick={[Function]}
  >
    2013
  </button>,
  <button
    className="filter-button "
    data-testid="yearFilter"
    onClick={[Function]}
  >
    2014
  </button>,
  <button
    className="filter-button "
    data-testid="yearFilter"
    onClick={[Function]}
  >
    2015
  </button>,
  <button
    className="filter-button "
    data-testid="yearFilter"
    onClick={[Function]}
  >
    2016
  </button>,
  <button
    className="filter-button "
    data-testid="yearFilter"
    onClick={[Function]}
  >
    2017
  </button>,
  <button
    className="filter-button "
    data-testid="yearFilter"
    onClick={[Function]}
  >
    2018
  </button>,
  <button
    className="filter-button "
    data-testid="yearFilter"
    onClick={[Function]}
  >
    2019
  </button>,
  <button
    className="filter-button "
    data-testid="yearFilter"
    onClick={[Function]}
  >
    2020
  </button>,
]
`;
